src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs

```diff
@@ -0,0 +1,373 @@
+
+    private int previousIntrumentStartIndex;
+
+    private int previousMetricPointStartIndex;
+
+    internal IMetricDataTransport MetricDataTransport;
+
+    public OtlpProtobufSerializer(IMetricDataTransport metricDataTransport)
+    {
+        this.MetricDataTransport = metricDataTransport;
+    }
+
+    internal void SerializeAndSendMetrics(byte[] buffer, ref int currentPosition, Resource resource, in Batch<Metric> metricBatch)
+    {
+        foreach (var metric in metricBatch)
+        {
+            if (this.scopeMetrics.TryGetValue(metric.MeterName, out var metricList))
+            {
+                metricList.Add(metric);
+            }
+            else
+            {
+                if (MetricListPool.TryTake(out var newList))
+                {
+                    newList.Add(metric);
+                    this.scopeMetrics[metric.MeterName] = newList;
+                }
+                else
+                {
+                    newList = new List<Metric>();
+                    newList.Add(metric);
+                    this.scopeMetrics[metric.MeterName] = newList;
+                }
+            }
+        }
+
+        // Serialize
+        this.SerializeResourceMetrics(buffer, ref currentPosition, resource);
+
+        this.ReturnScopeMetrics();
+    }
+
+    internal void ReturnScopeMetrics()
+    {
+        foreach (var list in this.scopeMetrics)
+        {
+            list.Value.Clear();
+            MetricListPool.Add(list.Value);
+        }
+
+        this.scopeMetrics.Clear();
+    }
+
+    internal void SerializeResourceMetrics(byte[] buffer, ref int currentPosition, Resource resource)
+    {
+        this.resourceMetricStartIndex = currentPosition;
+
+        currentPosition += LengthAndTagSize;
+
+        // Serialize Resource
+        // TODO: Avoid serializing it multiple times.
+        SerializeResource(buffer, ref currentPosition, resource);
+
+        // TODO: Serialize schema_url field
+
+        // Serialize ScopeMetrics field
+        foreach (KeyValuePair<string, List<Metric>> metric in this.scopeMetrics)
+        {
+            this.previousScopeMetricsStartIndex = currentPosition;
+
+            currentPosition += LengthAndTagSize;
+
+            // Serialize this meter/scope
+            this.SerializeScopeMetrics(buffer, ref currentPosition, metric.Key, metric.Value);
+
+            // Reset the current position to start writing again from where we started the last scope.
+            currentPosition = this.previousScopeMetricsStartIndex;
+        }
+    }
+
+    internal void SerializeScopeMetrics(byte[] buffer, ref int currentPosition, string scopeName, List<Metric> metrics)
+    {
+        bool scopeInformationIncluded = false;
+
+        foreach (Metric metric in metrics)
+        {
+            if (!scopeInformationIncluded)
+            {
+                // Serialize scope information
+                SerializeInstrumentationScope(buffer, ref currentPosition, scopeName, metric.MeterTags);
```

Comment:
Could you move this outside of the foreach loop? This method would only be called for metrics of the same `Meter`.